### ISSUE TITLE
Alternative UI implementations (long-term)

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#708]
I think we ought not to do this just yet, but part of the long term plan is to have different UIs be able to run Image Sequencer. I'm collecting a few ideas all in one place, but we have plenty of work before these seem to be a reasonable priority to take up. 

 * "node-and-line" style UI - like a flow-based program (Quartz Composer, Pure Data, node red (javascript!), etc)
    * visual nodes-and-lines UI: https://github.com/flowhub/the-graph
    * https://flowhub.github.io/the-graph/examples/demo-simple.html
* Possible web-based commandline interface: https://hyper.is/?
* (for node-and-line style UIs) non-linear sequences with Y-splitters -- ability to split an image into two sequences! (whoa) 

(Many from https://github.com/publiclab/image-sequencer/issues/118)

**Comment by jywarren** [comment]
More!

* https://developers.google.com/blockly/
* Materialize?

**Comment by HarshKhandeparkar** [comment]
@jywarren I'll try to change some basic looks using materialize. I have a bit of experience with that

**Comment by HarshKhandeparkar** [comment]
@jywarren @publiclab/is-reviewers I just had a great idea in mind. How about making an image-sequencer desktop app? Using electron we can create that app. It will be cross platform and very easy to build too since we can easily copy the current website code and run it in electron. It can be a good example of ui implementation. It can be in a separate repository. Thoughts?

**Comment by HarshKhandeparkar** [comment]
@Divy123 @jywarren how about creating a new repo for different image-sequencer ui implementations with reactjs, blockly etc. While keeping the legacy ui in the current repo but planning a complete overhaul for that(stylewise only) because purejs/html is better for first timers. Thoughts?

**Comment by VibhorCodecianGupta** [comment]
@jywarren I agree with @Divy123 about using frameworks for the long run. The UI is very light for now and pure JS and HTML are just about perfect to render it, but should the project UI get an overhaul with a very impressive UI/UX structure, we might need to look into frameworks to make our jobs easier and make the code more legible

**Comment by HarshKhandeparkar** [comment]
@VibhorCodecianGupta the UI overhaul can be done stylewise by using some
modern CSS framework like materialize to make the page look modern and flat
and then the same UI can be copied over to react and just break the
different parts of the page into different components. Thoughts?

**Comment by Divy123** [comment]
@HarshKhandeparkar agreed but there is more to it.
Since as of now we are dealing with lots of UI states I think working with redux can be a great advantage.
After employing react and redux together e can surely work on styles or rather we can work simultaneously.

**Comment by HarshKhandeparkar** [comment]
@Divy123 you can work on React, I will work on styles. We can then merge both maybe?

**Comment by Divy123** [comment]
Sure @HarshKhandeparkar that would be great.
Lets know @jywarren thinks upon this.

**Comment by HarshKhandeparkar** [comment]
@jywarren @Divy123 @publiclab/is-reviewers How does this look?

![new-select](https://user-images.githubusercontent.com/34770591/51754285-e6989300-20e1-11e9-9307-fa74a32eba55.gif)


![new-range](https://user-images.githubusercontent.com/34770591/51754297-ed270a80-20e1-11e9-8883-4e6a31a3b754.gif)


![new-inputs](https://user-images.githubusercontent.com/34770591/51754304-f2845500-20e1-11e9-9b68-aa523121aa3d.gif)

**Comment by Divy123** [comment]
That's super cool!!
Awesome !!

**Comment by HarshKhandeparkar** [comment]
> That's super cool!!
> Awesome !!
> While right now I am struggling with my npm .
> The npm install is too slow

Is that because of internet or something else?

**Comment by Divy123** [comment]
Actually it was due to internet.

**Comment by HarshKhandeparkar** [comment]
> Actually it was due to internet.

Where do you live?

**Comment by Divy123** [comment]
Allahabad or rather Prayagraj :laughing:

**Comment by HarshKhandeparkar** [comment]
> Allahabad or rather Prayagraj

😁

**Comment by HarshKhandeparkar** [comment]
Were you a gci student by the way😅?

**Comment by Divy123** [comment]
@HarshKhandeparkar is there something i can contact you personally?

**Comment by HarshKhandeparkar** [comment]
> @HarshKhandeparkar is there something i can contact you personally?

I'm a minor and I rather like to be private. Sorry

**Comment by Divy123** [comment]
No I wasn't.

**Comment by HarshKhandeparkar** [comment]
> No I wasn't.

Ok 😅

**Comment by Divy123** [comment]
Sure @HarshKhandeparkar . Also i think we should delete these comments to make the issue discussion cleaner.

**Comment by HarshKhandeparkar** [comment]
Ya sure lets do that

**Comment by HarshKhandeparkar** [comment]
Collapsible update! @jywarren @Divy123 @publiclab/is-reviewers 


![collapsible-new](https://user-images.githubusercontent.com/34770591/51787355-c70f7200-2196-11e9-9241-0889ed310b0e.gif)

**Comment by Divy123** [comment]
I think #670 can be a part if we plan to use React and redux.

**Comment by HarshKhandeparkar** [comment]
Update!!! @publiclab/is-reviewers @jywarren 

Download! :


![new-download](https://user-images.githubusercontent.com/34770591/51824479-64d67e80-2308-11e9-9e89-fec35455211e.gif)


Insert Step! :


![new-insert](https://user-images.githubusercontent.com/34770591/51824521-7ddf2f80-2308-11e9-9aa1-74505469234b.gif)


If it looks like the animations are a bit stuttery, they are not. Its because the GIF frame rate is low. Thanks!

**Comment by HarshKhandeparkar** [comment]
@publiclab/is-reviewers @Divy123 @jywarren Update!!

new header:

![new-header](https://user-images.githubusercontent.com/34770591/51923730-938c4c00-2411-11e9-8c57-c84749de9375.gif)


new footer:

![new-footer](https://user-images.githubusercontent.com/34770591/51923745-98e99680-2411-11e9-936a-bceb079cb91f.gif)


new range:

![no-span-range](https://user-images.githubusercontent.com/34770591/51923779-a7d04900-2411-11e9-988b-71c3cfd697ca.gif)


new responsive page design:

![responsive-divs](https://user-images.githubusercontent.com/34770591/51923794-b0c11a80-2411-11e9-9002-67c8f4f1a897.gif)



new responsive step images:

![responsive-step-img](https://user-images.githubusercontent.com/34770591/51923807-b7e82880-2411-11e9-8854-753758c52c9f.gif)



new self-adjusting textareas instead of inputs:

![responsive-textareas](https://user-images.githubusercontent.com/34770591/51923829-c46c8100-2411-11e9-96ed-8779a49e6c20.gif)

**Comment by Divy123** [comment]
Super awesome. Looks amazing.
@jywarren please have a look and also I want to migrate the code to React and redux.
Should I proceed ahead?

**Comment by HarshKhandeparkar** [comment]
@Divy123 I have made some other general changes to the code as well. Please have a look at it if you haven't. Thanks!

https://github.com/HarshKhandeparkar/image-sequencer/commits/materialize-ui-overhaul

**Comment by jywarren** [comment]
I also love this 💕💕, but I really want to encourage folks to consider
letting our codebase mature a bit more before attempting a new UI. I know
it's exciting to try out but we have a lot of things to work on to flesh
out and stabilize the codebase before I think we're ready for a new UI.

Make no mistake, i truly appreciate your energy and enthusiasm here. But
let's look at a couple of the big projects we need to complete before
opening a new one.

For example, the idea of truly standalone meta modules #315, or full UI
testing, or per module comparison testing. And then of course the big
project of a serverless express app to allow cloud-based processing. Also
webassembly acceleration! Many are listed at
https://github.com/publiclab/image-sequencer/labels/project

These are all critical projects and once we build them out we'll be in a
much better place to take on a new big project because the core of our
library will be stronger.

I'm 100% in support of this new UI, but I want to encourage us to think
about it in the /next/ cycle of big projects, rather than right now. I hope
that sounds reasonable and fair!!! What do you all think?

And once more, this is really impressive!!!!

**Comment by HarshKhandeparkar** [comment]
@jywarren Materialize is a very similar to bootstrap and even a bit simpler. It has no dependencies either. I have only used materialize instead of bootstrap. I have also fixed errors and redundancies in the index.html file. I have also tried to minimise custom styles. Also I have changed the js code and split it into two different folders one for DOM and one for sequencer actions. Also fixed a few code errors. All of the code is almost the same with a few changes, mostly changes in classes. I have replaced the bootstrap components with similar looking materialize components. The end result is that the UI looks better since equivalent materialize components look better than bootstrap. Also the inputs have been changed to reduce custom code and to increase esse of input(in case of textarea)

**Comment by Divy123** [comment]
Sure @jywarren .
Also I think the changes made by @HarshKhandeparkar are easily adoptable .
The idea of using React can be postponed until the changes as said by you are done.

**Comment by jywarren** [comment]
Ok, well Harsh maybe share a PR, but I worry about the changes affecting
the display of button state for many custom UI behaviors, and a number of
UIs relied on bootstrap. Let's give it a good testing but I hope you
understand my caution in this!

You can apart from opening a PR also push to your own gh-pages branch to
let other people try it out. Thanks!

**Comment by HarshKhandeparkar** [comment]
@jywarren sure. I think I will open a PR and any you can point out any concerning things? This is still a WIP though

**Comment by HarshKhandeparkar** [comment]
@jywarren do my gh-pages link to `harshkhandepsrkar.github.io`? Or some other link? I haven't tried it before.

**Comment by HarshKhandeparkar** [comment]
@jywarren yess! That's a great idea. I am going to open a pr right now. Thanks for the help with gh-pages.

**Comment by Harshithpabbati** [comment]
@HarshKhandeparkar a great work!! :+1:

**Comment by Harshithpabbati** [comment]
May be few things need to be added to it.Fine will make them as issues after this get's merged

**Comment by HarshKhandeparkar** [comment]
Thanks. I'm actually going to refactor the ui code by separating the sequencer part of it from the DOM part and I'm working on it [here](https://github.com/publiclab/image-sequencer/blob/main/src/Modules.js). This links to a branch in my fork. Please check out the `ui/` folder and `examples/` folder.

**Comment by Harshithpabbati** [comment]
@jywarren we can make a big move by using react . I am interested in doing it.

**Comment by Divy123** [comment]
I am also interested.
May be we can collaborate?

**Comment by Harshithpabbati** [comment]
Yeah we can once @jywarren agrees with it :smile:

**Comment by Harshithpabbati** [comment]
@jywarren how about making it into a dynamic site we can have login system and once he get's into his account he can have all his data about the saved sequence and the images which he want for future use. What are your thought about it??

**Comment by Harshithpabbati** [comment]
@jywarren we need your thoughts here.

**Comment by 28-ys** [comment]
@harshithpabbati  nice idea about making it into a dynamic site.
